stage0/src/image_header.rs

```diff
@@ -2,155 +2,106 @@
-        return None;
-    }
-
-    if !imagea.check_bounds(sig_addr + 4 + sig_size) {
-        return None;
-    }
+    let img = Image(imagea);
 
-    // Check what is supposed to be the full image length
-    if !imagea.check_bounds(img_start + table.total_image_len) {
+    if !img.validate() {
         return None;
     }
 
-    // We can now say that the following are safe
-    // - Accessing the full image range
-    // - Accessing the key range
-    // - Accessing the signature range
-    Some(imagea)
-}
-
-// The careful observer will note that yes this is just the
-// start of an ARMv8m image with extra data shoved in the
-// vector table
-#[repr(C)]
-pub struct ImageHeader {
-    sp: u32,
-    pc: u32,
-    _vector_table: [u8; 24],
-    image_length: u32,
-    _image_type: u32,
-    header_offset: u32,
+    Some(img)
 }
 
-impl ImageHeader {
-    pub fn get_img_start(&self) -> u32 {
-        self as *const Self as u32
-    }
-
-    fn check_bounds(&self, address: u32) -> bool {
-        let start = self.get_img_start();
-
-        return address >= start && address < (start + self.image_length);
+impl Image {
+    fn get_img_start(&self) -> u32 {
+        self.0 as *const ImageVectors as u32
     }
 
-    fn get_table_start(&self) -> u32 {
-        self.get_img_start() + self.header_offset
+    fn get_header(&self) -> *const ImageHeader {
+        // SAFETY: This generated by the linker script which we trust
+        // Note that this is generated from _this_ image's linker script
+        // as opposed to the _image_ linker script but those two _must_
+        // be the same value!
+        let vector_size = unsafe { core::ptr::addr_of!(__vector_size) as u32 };
+        (self.get_img_start() + vector_size) as *const ImageHeader
     }
 
     /// Make sure all of the image flash is programmed
-    fn validate(&self) -> bool {
+    pub fn validate(&self) -> bool {
         let img_start = self.get_img_start();
 
-        // Start by making sure the region is actually programmed
+        // Start by making sure we can access the page where the vectors live
         let valid = lpc55_romapi::validate_programmed(img_start, 0x200);
 
         if !valid {
             return false;
         }
 
+        let header_ptr = self.get_header();
+
+        // Next validate the header location is programmed
+        let valid = lpc55_romapi::validate_programmed(header_ptr as u32, 0x200);
+
+        if !valid {
+            return false;
+        }
+
+        // SAFETY: We've validated the header location is programmed so this
+        // will not trigger a fault. This is generated from our build scripts
+        // which we trust.
+        let header = unsafe { &*header_ptr };
+
         // Next make sure the marked image length is programmed
         let valid = lpc55_romapi::validate_programmed(
             img_start,
-            (self.image_length + 0x1ff) & !(0x1ff),
+            (header.total_image_len + 0x1ff) & !(0x1ff),
```

Comment:
Nit: these `0x200` and `0x1ff`s -- are these the programming page size? A named constant might be good.